libdnf/advisory/advisory.cpp

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright (C) 2021 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "libdnf/advisory/advisory.hpp"
+
+#include "libdnf/advisory/advisory_collection.hpp"
+#include "libdnf/advisory/advisory_reference.hpp"
+#include "libdnf/rpm/solv_sack_impl.hpp"
+
+#include <solv/pool.h>
+
+namespace libdnf::advisory {
+
+Advisory::Advisory(libdnf::Base & base, AdvisoryId id) : id(id), base(base) {}
+
+std::string Advisory::get_name() const {
+    const char * name;
+    auto & solv_sack = base.get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+
+    name = pool_lookup_str(pool, id.id, SOLVABLE_NAME);
+    size_t prefix_len = strlen("patch:");
+
+    assert(strncmp("patch:", name, prefix_len) == 0);
+
+    return std::string(name + prefix_len);
+}
+
+Advisory::Type Advisory::get_type() const {
+    const char * type;
+    auto & solv_sack = base.get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+    type = pool_lookup_str(pool, id.id, SOLVABLE_PATCHCATEGORY);
+
+    if (type == NULL) {
+        return Advisory::Type::UNKNOWN;
+    }
+    if (!strcmp(type, "bugfix")) {
+        return Advisory::Type::BUGFIX;
+    }
+    if (!strcmp(type, "enhancement")) {
+        return Advisory::Type::ENHANCEMENT;
+    }
+    if (!strcmp(type, "security")) {
+        return Advisory::Type::SECURITY;
+    }
+    if (!strcmp(type, "newpackage")) {
+        return Advisory::Type::NEWPACKAGE;
+    }
+
+    return Advisory::Type::UNKNOWN;
+}
+
+const char * Advisory::advisory_type_to_cstring(Type type) {
+    switch (type) {
+        case Type::ENHANCEMENT:
+            return "enhancement";
+        case Type::SECURITY:
+            return "security";
+        case Type::NEWPACKAGE:
+            return "newpackage";
+        case Type::BUGFIX:
+            return "bugfix";
+        default:
+            return NULL;
+    }
+}
+
+std::string Advisory::get_severity() const {
+    auto & solv_sack = base.get_rpm_solv_sack();
+    Pool * pool = solv_sack.p_impl->get_pool();
+
+    return std::string(pool_lookup_str(pool, id.id, UPDATE_SEVERITY));
```

Comment:
I am not really sure how the internalization works or what it even does but I saw how it is used with `SolvQuery` and `Package` in solv/package_private. In subsequent PRs I would like to do something similar for advisories (solv/adviosry_private) because they have the same problem with either duplicating logic or constructing expensive `Advisory` object in `AdvisoryQuery`.

Can we address this later? in the mentioned solv/advisory_private?

Reply:
`solv/adviosry_private` is good idea and it can be delivered in another PR.
Also I would like to propose a minor change in the code. What about to replace the return line with
```
auto severity = pool_lookup_str(pool, id.id, UPDATE_SEVERITY);
return severity ? std::string(severity) : std::string();
```

Reply:
Or you can keep it like it is, because there is TODO